service/history/workflow/mutable_state_impl.go

```diff
@@ -5906,7 +5907,70 @@ func (ms *MutableStateImpl) closeTransactionTrackTombstones(
 		return
 	}
 
+	// Get current versioned transition
+	transitionHistory := ms.executionInfo.TransitionHistory
+	currentVersionedTransition := transitionHistory[len(transitionHistory)-1]
+
+	// Track HSM node deletions
 	var tombstones []*persistencespb.StateMachineTombstone
+	if ms.stateMachineNode != nil {
+		opLog, err := ms.stateMachineNode.Outputs()
+		if err != nil {
+			ms.logger.Error("Failed to get HSM operation log", tag.Error(err))
+			return
+		}
+
+		for _, op := range opLog {
+			if deleteOp, ok := op.(hsm.DeleteOperation); ok {
+				path := deleteOp.Path()
+				if len(path) == 0 {
+					continue // Skip root deletion
+				}
+				key := path[len(path)-1]
+
+				// Create appropriate tombstone type based on path/key
+				var tombstone *persistencespb.StateMachineTombstone
+				switch key.Type {
+				case "Activity":
+					tombstone = &persistencespb.StateMachineTombstone{
+						StateMachineKey: &persistencespb.StateMachineTombstone_ActivityScheduledEventId{
+							ActivityScheduledEventId: ms.parseID(key.ID),
+						},
+					}
+				case "Timer":
+					tombstone = &persistencespb.StateMachineTombstone{
+						StateMachineKey: &persistencespb.StateMachineTombstone_TimerId{
+							TimerId: key.ID,
+						},
+					}
+				case "ChildExecution":
+					tombstone = &persistencespb.StateMachineTombstone{
+						StateMachineKey: &persistencespb.StateMachineTombstone_ChildExecutionInitiatedEventId{
+							ChildExecutionInitiatedEventId: ms.parseID(key.ID),
+						},
+					}
+				case "RequestCancel":
+					tombstone = &persistencespb.StateMachineTombstone{
+						StateMachineKey: &persistencespb.StateMachineTombstone_RequestCancelInitiatedEventId{
+							RequestCancelInitiatedEventId: ms.parseID(key.ID),
+						},
+					}
+				case "Signal":
+					tombstone = &persistencespb.StateMachineTombstone{
+						StateMachineKey: &persistencespb.StateMachineTombstone_SignalExternalInitiatedEventId{
+							SignalExternalInitiatedEventId: ms.parseID(key.ID),
+						},
+					}
+				}
```

Comment:
This is not necessary, you only need to track `StateMachinePath` changes here. Activities and other state machines pre-date the HSM framework and won't be reflected in the oplog.

Reply:
Changed! Thanks